src/shogun/classifier/AveragedPerceptron.cpp

```diff
@@ -103,14 +87,8 @@ bool CAveragedPerceptron::train_machine(CFeatures* data)
 		SG_WARNING("Averaged Perceptron algorithm did not converge after %d iterations.\n", max_iter)
 
 	// calculate and set the average paramter of w, bias
-	for (int32_t i=0; i<num_feat; i++)
-		w[i]=tmp_w[i]/(num_vec*iter);
+	linalg::scale(tmp_w, w, 1.0 / (num_vec * iter));
 	bias=tmp_bias/(num_vec*iter);
 
-	SG_FREE(output);
-	SG_FREE(tmp_w);
-
 	set_w(w);
```

Comment:
i would prefer to remove this and udpate the state vector itself inside the main loop.
@shubham808 can elaborate as he is doing similar stuff for the perceptron

Reply:
@karlnapf I think `set_w` has been already called inside the iteration function?